Build without type and topic discovery

Since pull request #294 was merged, I cannot build with `cmake .. -DENABLE_TYPE_DISCOVERY=OFF -DENABLE_TOPIC_DISCOVERY=OFF`:

```
/cyclonedds-cxx/src/ddscxx/include/dds/topic/detail/TTopicImpl.hpp: In instantiation of ‘static dds::topic::Topic<T, dds::topic::detail::Topic> dds::topic::detail::Topic<T>::discover_topic(const DomainParticipant&, const string&, const dds::core::Duration&) [with T = dds::topic::TSubscriptionBuiltinTopicData<org::eclipse::cyclonedds::topic::SubscriptionBuiltinTopicDataDelegate>; dds::domain::DomainParticipant = dds::domain::TDomainParticipant<org::eclipse::cyclonedds::domain::DomainParticipantDelegate>; std::__cxx11::string = std::__cxx11::basic_string<char>]’:
/cyclonedds-cxx/src/ddscxx/include/org/eclipse/cyclonedds/topic/discovery.hpp:85:60:   required from ‘static dds::topic::Topic<T, dds::topic::detail::Topic> org::eclipse::cyclonedds::topic::typed_lookup_topic<T, dds::topic::detail::Topic<T> >::discover(const DomainParticipant&, const string&, const dds::core::Duration&) [with T = dds::topic::TSubscriptionBuiltinTopicData<org::eclipse::cyclonedds::topic::SubscriptionBuiltinTopicDataDelegate>; dds::domain::DomainParticipant = dds::domain::TDomainParticipant<org::eclipse::cyclonedds::domain::DomainParticipantDelegate>; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/cyclonedds-cxx/src/ddscxx/include/org/eclipse/cyclonedds/topic/discovery.hpp:107:96:   required from ‘static TOPIC org::eclipse::cyclonedds::topic::lookup_topic<TOPIC, DELEGATE>::discover(const DomainParticipant&, const string&, const dds::core::Duration&) [with TOPIC = dds::topic::Topic<dds::topic::TSubscriptionBuiltinTopicData<org::eclipse::cyclonedds::topic::SubscriptionBuiltinTopicDataDelegate>, dds::topic::detail::Topic>; DELEGATE = dds::topic::detail::Topic<dds::topic::TSubscriptionBuiltinTopicData<org::eclipse::cyclonedds::topic::SubscriptionBuiltinTopicDataDelegate> >; dds::domain::DomainParticipant = dds::domain::TDomainParticipant<org::eclipse::cyclonedds::domain::DomainParticipantDelegate>; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/cyclonedds-cxx/src/ddscxx/include/dds/topic/detail/discovery.hpp:40:105:   required from ‘TOPIC dds::topic::discover(const DomainParticipant&, const string&, const dds::core::Duration&) [with TOPIC = dds::topic::Topic<dds::topic::TSubscriptionBuiltinTopicData<org::eclipse::cyclonedds::topic::SubscriptionBuiltinTopicDataDelegate>, dds::topic::detail::Topic>; dds::domain::DomainParticipant = dds::domain::TDomainParticipant<org::eclipse::cyclonedds::domain::DomainParticipantDelegate>; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/cyclonedds-cxx/src/ddscxx/src/org/eclipse/cyclonedds/sub/BuiltinSubscriberDelegate.cpp:118:61:   required from ‘dds::sub::DataReader<T, dds::sub::detail::DataReader> org::eclipse::cyclonedds::sub::create_builtin_reader(org::eclipse::cyclonedds::sub::SubscriberDelegate&, const string&) [with T = dds::topic::TSubscriptionBuiltinTopicData<org::eclipse::cyclonedds::topic::SubscriptionBuiltinTopicDataDelegate>; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/cyclonedds-cxx/src/ddscxx/src/org/eclipse/cyclonedds/sub/BuiltinSubscriberDelegate.cpp:184:112:   required from here
/cyclonedds-cxx/src/ddscxx/include/dds/topic/detail/TTopicImpl.hpp:312:77: error: ‘getTypeInfo’ is not a member of ‘org::eclipse::cyclonedds::topic::TopicTraits<dds::topic::TSubscriptionBuiltinTopicData<org::eclipse::cyclonedds::topic::SubscriptionBuiltinTopicDataDelegate> >’
src/ddscxx/CMakeFiles/ddscxx.dir/build.make:666: recipe for target 'src/ddscxx/CMakeFiles/ddscxx.dir/src/org/eclipse/cyclonedds/sub/BuiltinSubscriberDelegate.cpp.o' failed
make[2]: *** [src/ddscxx/CMakeFiles/ddscxx.dir/src/org/eclipse/cyclonedds/sub/BuiltinSubscriberDelegate.cpp.o] Error 1
CMakeFiles/Makefile2:199: recipe for target 'src/ddscxx/CMakeFiles/ddscxx.dir/all' failed
make[1]: *** [src/ddscxx/CMakeFiles/ddscxx.dir/all] Error 2
Makefile:170: recipe for target 'all' failed
make: *** [all] Error 2
``` 
Several errors of that sort (getTypeInfo), posted only one for brevity. 

TBH I don't understand the change well enough to know if Type and Topic discovery are mandatory options now and if the cmake options should be removed or some fix is required. 

Last commit that works for me is ce6d5e6 , Ubuntu 18.04. 

I discovered this issue when building [cyclonedds with CMake 3.10](https://github.com/eclipse-cyclonedds/cyclonedds/issues/1354), as then this project won't pick up the property and will attempt to build without discovery.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -9,7 +9,7 @@
 #
 # SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
 #
-cmake_minimum_required(VERSION 3.10)
+cmake_minimum_required(VERSION 3.16)
 project(CycloneDDS-CXX VERSION 0.10.0 LANGUAGES C CXX)
 
 set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake/Modules")
@@ -108,12 +108,6 @@ if(CMAKE_GENERATOR STREQUAL "Xcode")
   set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_UNUSED_VARIABLE YES)
   set(CMAKE_XCODE_ATTRIBUTE_CLANG_WARN_DOCUMENTATION_COMMENTS YES)
   set(CMAKE_XCODE_ATTRIBUTE_GCC_WARN_ABOUT_MISSING_PROTOTYPES YES)
-endif()
-
-if(CMAKE_VERSION VERSION_LESS 3.13)
-  macro(add_link_options)
-    link_libraries(${ARGV})
-  endmacro()
 endif()
 
 if(CMAKE_CROSSCOMPILING)

--- a/README.md
+++ b/README.md
@@ -30,7 +30,7 @@ following installed on your host:
  * C and C++ compilers (most commonly GCC on Linux, Visual Studio on Windows,
    Xcode on macOS);
  * [Git](https://git-scm.com/) version control system;
- * [CMake](https://cmake.org/download/), version 3.10 or later;
+ * [CMake](https://cmake.org/download/), version 3.16 or later;
  * [Eclipse Cyclone DDS](https://github.com/eclipse-cyclonedds/cyclonedds/)
 
 *Eclipse Cyclone DDS* has dependencies of its own, most notably Bison. To

--- a/examples/helloworld/CMakeLists.txt
+++ b/examples/helloworld/CMakeLists.txt
@@ -10,7 +10,7 @@
 # SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
 #
 project(helloworld LANGUAGES C CXX)
-cmake_minimum_required(VERSION 3.5)
+cmake_minimum_required(VERSION 3.16)
 
 if(NOT TARGET CycloneDDS-CXX::ddscxx)
   find_package(CycloneDDS-CXX REQUIRED)

